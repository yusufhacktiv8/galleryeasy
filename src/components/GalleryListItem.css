.gallery-list-item {
  height: 160px;
  min-width: 222px;
  overflow: hidden;
  position: relative;
  top: 0;
  left: 0;
  background-color: black;
}

.tag-img {
  float: left;
  position: absolute;
  top: 115px;
  left: 170px;
  cursor: pointer;
}

.tag-img-transparent {
  opacity: 0.6;
  filter: alpha(opacity=60);
}
