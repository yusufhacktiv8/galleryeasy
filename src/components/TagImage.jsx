import React from 'react';

const TAG_IMAGE_URL = '/images/tagimage.png';

export default ({ transparent, onClick }) => {
  const className = transparent ? 'img-transparent-50' : '';
  return (
    <img src={TAG_IMAGE_URL} className={className} onClick={onClick} alt="Tag" />
  );
};
