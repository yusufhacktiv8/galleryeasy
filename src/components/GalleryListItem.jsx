import React, { Component } from 'react';
import TagImage from './TagImage';

const Status = {
  DEFAULT: 'default',
  HOVERED: 'hovered',
};

export default class GalleryListItem extends Component {
  constructor(props) {
    super(props);
    this.state = {
      status: Status.DEFAULT,
    };
  }

  onMouseEnter = () => {
    this.setState({
      status: Status.HOVERED,
    });
  }

  onMouseOut = () => {
    this.setState({
      status: Status.DEFAULT,
    });
  }

  isHovered() {
    return this.state.status === Status.HOVERED;
  }

  render() {
    const { url, favourited } = this.props;
    let tagImage = null;
    if (!favourited && this.isHovered()) {
      tagImage = <TagImage transparent />;
    } else if (favourited) {
      tagImage = <TagImage />;
    }

    return (
      <div
        role="link"
        className="gallery-list-item"
        onMouseEnter={this.onMouseEnter}
        onMouseOut={this.onMouseOut}
      >
        <img url={url} />
        {tagImage}
      </div>
    );
  }
}
