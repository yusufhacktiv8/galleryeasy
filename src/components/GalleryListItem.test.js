import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import sinon from 'sinon';

import GalleryListItem from './GalleryListItem';
import TagImage from './TagImage';

test('Default GalleryListItem correctly rendered', () => {
  const component = renderer.create(
    <GalleryListItem url="/test.png" />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Not favourited GalleryListItem render transparent tag image when hovered', () => {
  const component = renderer.create(
    <GalleryListItem url="/test.png" />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  tree.props.onMouseEnter();
  
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Not favourited GalleryListItem remove transparent tag image when mouse out', () => {
  const component = renderer.create(
    <GalleryListItem url="/test.png" />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  tree.props.onMouseEnter();
  
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  
  tree.props.onMouseOut();
  
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Favourited GalleryListItem correctly rendered', () => {
  const component = renderer.create(
    <GalleryListItem url="/test.png" favourited />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Favourited GalleryListItem render tag image when hovered', () => {
  const component = renderer.create(
    <GalleryListItem url="/test.png" favourited />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  tree.props.onMouseEnter();
  
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Favourited GalleryListItem still render tag image when mouse out', () => {
  const component = renderer.create(
    <GalleryListItem url="/test.png" favourited />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  tree.props.onMouseEnter();
  
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  
  tree.props.onMouseOut();
  
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Not favourited GalleryListItem trigger onFavouriteClick when TagImage clicked', () => {
  const callback = sinon.spy();
  const favourited = false;
  const element = mount(<GalleryListItem url="/test.png" onFavouriteClicked={callback} />);
  element.simulate('mouseEnter');
  element.find(TagImage).simulate('click');
  
  expect(callback.calledWith(favourited)).toBe(true);
});

test('Favourited GalleryListItem trigger onFavouriteClick when TagImage clicked', () => {
  const callback = sinon.spy();
  const favourited = true;
  const element = mount(<GalleryListItem favourited url="/test.png" onFavouriteClicked={callback} />);
  
  element.find(TagImage).simulate('click');
  expect(callback.calledWith(favourited)).toBe(true);
  
  element.simulate('mouseEnter');
  
  element.find(TagImage).simulate('click');
  expect(callback.calledWith(favourited)).toBe(true);
});
