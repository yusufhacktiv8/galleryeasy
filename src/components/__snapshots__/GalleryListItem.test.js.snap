// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default GalleryListItem correctly rendered 1`] = `
<div
  className="gallery-list-item"
  onMouseEnter={[Function]}
  onMouseOut={[Function]}
  role="link"
>
  <img
    url="/test.png"
  />
</div>
`;

exports[`Favourited GalleryListItem correctly rendered 1`] = `
<div
  className="gallery-list-item"
  onMouseEnter={[Function]}
  onMouseOut={[Function]}
  role="link"
>
  <img
    url="/test.png"
  />
  <img
    alt="Tag"
    className=""
    onClick={undefined}
    src="/images/tagimage.png"
  />
</div>
`;

exports[`Favourited GalleryListItem render tag image when hovered 1`] = `
<div
  className="gallery-list-item"
  onMouseEnter={[Function]}
  onMouseOut={[Function]}
  role="link"
>
  <img
    url="/test.png"
  />
  <img
    alt="Tag"
    className=""
    onClick={undefined}
    src="/images/tagimage.png"
  />
</div>
`;

exports[`Favourited GalleryListItem render tag image when hovered 2`] = `
<div
  className="gallery-list-item"
  onMouseEnter={[Function]}
  onMouseOut={[Function]}
  role="link"
>
  <img
    url="/test.png"
  />
  <img
    alt="Tag"
    className=""
    onClick={undefined}
    src="/images/tagimage.png"
  />
</div>
`;

exports[`Favourited GalleryListItem still render tag image when mouse out 1`] = `
<div
  className="gallery-list-item"
  onMouseEnter={[Function]}
  onMouseOut={[Function]}
  role="link"
>
  <img
    url="/test.png"
  />
  <img
    alt="Tag"
    className=""
    onClick={undefined}
    src="/images/tagimage.png"
  />
</div>
`;

exports[`Favourited GalleryListItem still render tag image when mouse out 2`] = `
<div
  className="gallery-list-item"
  onMouseEnter={[Function]}
  onMouseOut={[Function]}
  role="link"
>
  <img
    url="/test.png"
  />
  <img
    alt="Tag"
    className=""
    onClick={undefined}
    src="/images/tagimage.png"
  />
</div>
`;

exports[`Favourited GalleryListItem still render tag image when mouse out 3`] = `
<div
  className="gallery-list-item"
  onMouseEnter={[Function]}
  onMouseOut={[Function]}
  role="link"
>
  <img
    url="/test.png"
  />
  <img
    alt="Tag"
    className=""
    onClick={undefined}
    src="/images/tagimage.png"
  />
</div>
`;

exports[`Not favourited GalleryListItem remove transparent tag image when mouse out 1`] = `
<div
  className="gallery-list-item"
  onMouseEnter={[Function]}
  onMouseOut={[Function]}
  role="link"
>
  <img
    url="/test.png"
  />
</div>
`;

exports[`Not favourited GalleryListItem remove transparent tag image when mouse out 2`] = `
<div
  className="gallery-list-item"
  onMouseEnter={[Function]}
  onMouseOut={[Function]}
  role="link"
>
  <img
    url="/test.png"
  />
  <img
    alt="Tag"
    className="img-transparent-50"
    onClick={undefined}
    src="/images/tagimage.png"
  />
</div>
`;

exports[`Not favourited GalleryListItem remove transparent tag image when mouse out 3`] = `
<div
  className="gallery-list-item"
  onMouseEnter={[Function]}
  onMouseOut={[Function]}
  role="link"
>
  <img
    url="/test.png"
  />
</div>
`;

exports[`Not favourited GalleryListItem render transparent tag image when hovered 1`] = `
<div
  className="gallery-list-item"
  onMouseEnter={[Function]}
  onMouseOut={[Function]}
  role="link"
>
  <img
    url="/test.png"
  />
</div>
`;

exports[`Not favourited GalleryListItem render transparent tag image when hovered 2`] = `
<div
  className="gallery-list-item"
  onMouseEnter={[Function]}
  onMouseOut={[Function]}
  role="link"
>
  <img
    url="/test.png"
  />
  <img
    alt="Tag"
    className="img-transparent-50"
    onClick={undefined}
    src="/images/tagimage.png"
  />
</div>
`;
